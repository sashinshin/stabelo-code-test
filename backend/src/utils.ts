import config from "./config";
const { FLOORS } = config;

export const moveElevator = (destinationFloor: number, elevatorPositions) => {
    if (destinationFloor > FLOORS || 0 > destinationFloor) {
        throw new Error("Invalid destination floor")
    }

    for (let iterations = 0; FLOORS > iterations; iterations++) {
        for (let elevator = 0; elevatorPositions.length > elevator; elevator++) {
            const above = elevatorPositions[elevator][destinationFloor + iterations];
            if (above) {
                return {
                    elevator,
                    iterations,
                    above: true,
                    floorWithElevator: destinationFloor + iterations
                }
            }
            const below = elevatorPositions[elevator][destinationFloor - iterations];
            if (below) {
                return {
                    elevator,
                    iterations,
                    above: false,
                    floorWithElevator: destinationFloor - iterations
                }
            }
        };
    };

    throw new Error("Too many iterations")
}
