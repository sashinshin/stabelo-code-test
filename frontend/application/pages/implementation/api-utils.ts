const apiCall = async<T>(
    endpoint: string,
    method: string,
    headers: HeadersInit,
    body?: string
): Promise<T> => {
    const response = await fetch(`http://localhost:3000/${endpoint}`, {
        method,
        headers,
        body,
    });
    const parsed = await response.json();

    if (response.status !== 200) {
        throw Error(parsed.message);
    };
    return parsed.message;
};

export const getInterfaceSpecs = async (): Promise<number[][]> => {
    const response = await apiCall<{ elevators: number[], floors: number[] }>(
        "api/init",
        "GET",
        { "accepts": "application/json" }
    );
    return [response.elevators, response.floors];
};

export const getElevatorPositions = async (): Promise<ElevatorPositions> => {
    const response = await apiCall<ElevatorPositions>(
        "api/positions",
        "GET",
        { "accepts": "application/json" }
    );

    return response;
};

export const moveElevator = async (destinationFloor: number): Promise<number> => {
    const response = await apiCall<number>(
        "api/positions",
        "PATCH",
        { "Content-Type": "application/json" },
        JSON.stringify({ destinationFloor }),
    );

    return response;
};