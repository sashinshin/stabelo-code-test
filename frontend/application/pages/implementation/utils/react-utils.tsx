import * as React from "react";
import * as css from "../ImplementationPage.module.scss";

import { getElevatorPositions, moveElevator } from "./api-utils"

const getFloorId = (elevator: number, floor: number) => `e${elevator}f${floor}`;

const generateFloorsJSX = (
    elevator: number,
    floors: number[]
): JSX.Element[] => (floors.map((floor) => (
    <div id={getFloorId(elevator, floor)}
        key={getFloorId(elevator, floor)}
        className={css.cell}>
        F{floor}
    </div>)));

export const createInterfaceJSX = (
    elevators: number[],
    floors: number[]
): JSX.Element[] => (elevators.map((elevator) => (
    <div id={(elevator).toString()}
        key={elevator}>
        {generateFloorsJSX(elevator, floors)}
    </div>)));

export const visualizeElevators = (elevatorPositions: ElevatorPositions): void => {
    elevatorPositions.forEach((elevators, elevator) => {
        elevators.forEach((elevatorOnFloor, floor) => {
            const floorElement = document.getElementById(getFloorId(elevator, floor));
            elevatorOnFloor
                ? floorElement.classList.add(css["cell--active"])
                : floorElement.classList.remove(css["cell--active"]);
        });
    });
};

export const poller = async (floorsToTravel: number): Promise<void> => {
    if (floorsToTravel > 0) {
        setTimeout(async () => {
            const elevatorsPositions = await getElevatorPositions();
            visualizeElevators(elevatorsPositions);
            poller(floorsToTravel - 1);
        }, 2000);
    }
};

export const onClick = async (
    maxFloors: string,
    destinationFloor: string,
    setErrorMessage: React.Dispatch<React.SetStateAction<string>>
): Promise<void> => {
    try {
        const destinationFloorInt = Number.parseInt(destinationFloor, 10);
        if (destinationFloorInt > Number.parseInt(maxFloors, 10) || 0 > destinationFloorInt) {
            throw new Error("Invalid destination floor");
        }
        const floorsToTravel = await moveElevator(destinationFloorInt);
        poller(floorsToTravel);
    } catch (error) {
        setErrorMessage(error.message as string);
        setTimeout(() => {
            setErrorMessage("");
        }, 2000);
    }
};