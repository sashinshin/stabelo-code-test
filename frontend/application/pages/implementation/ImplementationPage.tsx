import "regenerator-runtime/runtime.js";
import * as React from "react";
import * as css from "./ImplementationPage.module.scss";
import { getInterfaceSpecs, getElevatorPositions, moveElevator } from "./utils/api-utils";
// import { visualizeElevators, createInterfaceJSX } from "./utils/react-utils";
import { createInterfaceJSX, visualizeElevators, poller } from "./utils/react-utils";


// const getFloorId = (elevator: number, floor: number) => `e${elevator}f${floor}`;

// const generateFloorsJSX = (
//     elevator: number,
//     floors: number[]
// ): JSX.Element[] => (floors.map((floor) => (
//     <div id={getFloorId(elevator, floor)}
//         key={getFloorId(elevator, floor)}
//         className={css.cell}>
//         F{floor}
//     </div>)));

// export const createInterfaceJSX = (
//     elevators: number[],
//     floors: number[]
// ): JSX.Element[] => (elevators.map((elevator) => (
//     <div id={(elevator).toString()}
//         key={elevator}>
//         {generateFloorsJSX(elevator, floors)}
//     </div>)));

// const visualizeElevators = (elevatorPositions: ElevatorPositions): void => {
//     elevatorPositions.forEach((elevators, elevator) => {
//         elevators.forEach((elevatorOnFloor, floor) => {
//             const floorElement = document.getElementById(getFloorId(elevator, floor));
//             elevatorOnFloor
//                 ? floorElement.classList.add(css["cell--active"])
//                 : floorElement.classList.remove(css["cell--active"]);
//         });
//     });
// };

// const poller = async (floorsToTravel: number): Promise<void> => {
//     if (floorsToTravel > 0) {
//         setTimeout(async () => {
//             const elevatorsPositions = await getElevatorPositions();
//             visualizeElevators(elevatorsPositions);
//             poller(floorsToTravel - 1);
//         }, 2000);
//     }
// };

// const onClick = async (
//     maxFloors: string,
//     destinationFloor: string,
//     setErrorMessage: React.Dispatch<React.SetStateAction<string>>
// ): Promise<void> => {
//     try {
//         const destinationFloorInt = Number.parseInt(destinationFloor, 10);
//         if (destinationFloorInt > Number.parseInt(maxFloors, 10) || 0 > destinationFloorInt) {
//             throw new Error("Invalid destination floor");
//         }
//         const floorsToTravel = await moveElevator(destinationFloorInt);
//         poller(floorsToTravel);
//     } catch (error) {
//         setErrorMessage(error.message as string);
//         setTimeout(() => {
//             setErrorMessage("");
//         }, 2000);
//     }
// };

const ImplementationPage = (): JSX.Element => {
    const [destinationFloor, setDestinationFloor] = React.useState<string>("0");
    const [elevatorInterface, setElevatorInterface] = React.useState<JSX.Element[]>([]);
    const [maxFloors, setMaxFloors] = React.useState<string>("0");
    const [errorMessage, setErrorMessage] = React.useState<string>("");

    const onClick = async (): Promise<void> => {
        try {
            const destinationFloorInt = Number.parseInt(destinationFloor, 10);
            const maxFloorsInt = Number.parseInt(maxFloors, 10);
            if (destinationFloorInt > maxFloorsInt || 0 > destinationFloorInt) {
                throw new Error("Invalid destination floor");
            }
            const floorsToTravel = await moveElevator(destinationFloorInt);
            poller(floorsToTravel);
        } catch (error) {
            setErrorMessage(error.message as string);
            setTimeout(() => {
                setErrorMessage("");
            }, 2000);
        }
    };

    React.useEffect((): void => {
        const init = async (): Promise<void> => {
            const [elevatorsArray, floorsArray, floors] = await getInterfaceSpecs();
            setMaxFloors(floors);
            setElevatorInterface(createInterfaceJSX(elevatorsArray, floorsArray));
            const elevatorsPositions = await getElevatorPositions();
            visualizeElevators(elevatorsPositions);
        }
        init();
    }, []);

    return (
        <>
            <div className={css["cell-container"]}>
                {elevatorInterface}
            </div>

            <div className={css["button-container"]}>
                <input
                    type="number"
                    min="0"
                    max={maxFloors}
                    value={destinationFloor}
                    onChange={(e) => setDestinationFloor(e.target.value)}
                />

                {errorMessage
                    ? <div>{errorMessage}</div>
                    : <button onClick={onClick}>Elevator</button>}

            </div>

        </>
    );
};

export default ImplementationPage;
