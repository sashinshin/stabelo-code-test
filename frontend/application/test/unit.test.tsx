import nock = require("nock");
import { enableFetchMocks } from 'jest-fetch-mock';
enableFetchMocks();
const mockedFetch = fetch as jest.Mocked<any>

// import fetchMock = require('jest-fetch-mock');
// fetchMock.enableFetchMocks()
import * as React from "react";
import * as renderer from "react-test-renderer";
import ImplementationPage from "../pages/implementation/ImplementationPage";


import { createInterfaceJSX } from "../pages/implementation/utils/react-utils"
import { getInterfaceSpecs, getElevatorPositions } from "../pages/implementation/utils/api-utils"

const test = () => {
    console.log(React);
    console.log(getElevatorPositions);
}
console.log(test);


const [FLOORS, ELEVATORS] = [5, 20];

export const generateElevatorPositions = (): ElevatorPositions => {
    const floors = new Array(FLOORS).fill(false);
    floors[0] = true;
    const elevators = new Array(ELEVATORS).fill(undefined).map(() => [...floors]);
    return elevators;
};

nock("http://localhost:3000/api")
    .get("/init")
    .reply(200, {
        message: {
            floorsArray: Array.from(Array(FLOORS).keys()).reverse(),
            elevatorsArray: Array.from(Array(ELEVATORS).keys()),
            floors: (FLOORS - 1).toString(),
        }
    });
nock("http://localhost:3000/api")
    .get("/positions")
    .reply(200, {
        message: generateElevatorPositions()
    });
nock("http://localhost:3000/api")
    .patch("/positions")
    .reply(200, {
        message: 2
    }
    );

describe("1.0 ImplementationPage test suite", () => {
    it("1.1 Mount ImplementationPage component", async () => {
        // given
        const mockedInitResponse = JSON.stringify({
            message: {
                floorsArray: Array.from(Array(FLOORS).keys()).reverse(),
                elevatorsArray: Array.from(Array(ELEVATORS).keys()),
            }
        });
        const mockedPositionsResponse = JSON.stringify({
            message: generateElevatorPositions()
        });
        mockedFetch
            .once(mockedInitResponse)
            .once(mockedPositionsResponse);

        // when
        const tree = renderer
            .create(<ImplementationPage />)
            .toJSON();

        // then
        expect(tree).toMatchSnapshot()
    });

    it("1.2 Generate interface JSX", async () => {
        // given
        const mockedResponse = JSON.stringify({
            message: {
                floorsArray: Array.from(Array(FLOORS).keys()).reverse(),
                elevatorsArray: Array.from(Array(ELEVATORS).keys()),
            }
        })
        mockedFetch.mockResponse(mockedResponse);

        // when
        const [elevatorsArray, floorsArray] = await getInterfaceSpecs();
        const interfaceJSX = createInterfaceJSX(elevatorsArray, floorsArray);

        // then
        expect(interfaceJSX).toMatchSnapshot();


    });
});

