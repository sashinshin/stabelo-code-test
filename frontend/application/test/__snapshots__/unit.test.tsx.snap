// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1.0 ImplementationPage test suite 1.1 Mount ImplementationPage component 1`] = `
Array [
  <div
    className="cell-container"
  />,
  <div
    className="button-container"
  >
    <input
      max="0"
      min="0"
      onChange={[Function]}
      type="number"
      value="0"
    />
    <button
      onClick={[Function]}
    >
      Elevator
    </button>
  </div>,
]
`;

exports[`1.0 ImplementationPage test suite 1.2 Generate interface JSX 1`] = `
Array [
  <div
    id="0"
  >
    <div
      className="cell"
      id="e0f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e0f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e0f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e0f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e0f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="1"
  >
    <div
      className="cell"
      id="e1f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e1f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e1f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e1f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e1f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="2"
  >
    <div
      className="cell"
      id="e2f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e2f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e2f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e2f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e2f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="3"
  >
    <div
      className="cell"
      id="e3f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e3f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e3f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e3f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e3f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="4"
  >
    <div
      className="cell"
      id="e4f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e4f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e4f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e4f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e4f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="5"
  >
    <div
      className="cell"
      id="e5f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e5f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e5f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e5f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e5f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="6"
  >
    <div
      className="cell"
      id="e6f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e6f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e6f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e6f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e6f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="7"
  >
    <div
      className="cell"
      id="e7f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e7f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e7f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e7f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e7f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="8"
  >
    <div
      className="cell"
      id="e8f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e8f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e8f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e8f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e8f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="9"
  >
    <div
      className="cell"
      id="e9f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e9f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e9f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e9f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e9f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="10"
  >
    <div
      className="cell"
      id="e10f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e10f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e10f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e10f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e10f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="11"
  >
    <div
      className="cell"
      id="e11f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e11f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e11f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e11f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e11f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="12"
  >
    <div
      className="cell"
      id="e12f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e12f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e12f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e12f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e12f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="13"
  >
    <div
      className="cell"
      id="e13f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e13f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e13f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e13f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e13f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="14"
  >
    <div
      className="cell"
      id="e14f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e14f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e14f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e14f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e14f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="15"
  >
    <div
      className="cell"
      id="e15f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e15f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e15f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e15f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e15f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="16"
  >
    <div
      className="cell"
      id="e16f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e16f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e16f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e16f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e16f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="17"
  >
    <div
      className="cell"
      id="e17f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e17f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e17f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e17f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e17f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="18"
  >
    <div
      className="cell"
      id="e18f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e18f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e18f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e18f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e18f0"
    >
      F
      0
    </div>
  </div>,
  <div
    id="19"
  >
    <div
      className="cell"
      id="e19f4"
    >
      F
      4
    </div>
    <div
      className="cell"
      id="e19f3"
    >
      F
      3
    </div>
    <div
      className="cell"
      id="e19f2"
    >
      F
      2
    </div>
    <div
      className="cell"
      id="e19f1"
    >
      F
      1
    </div>
    <div
      className="cell"
      id="e19f0"
    >
      F
      0
    </div>
  </div>,
]
`;
